File: apps/kotlin/src/test/kotlin/chapter10/FeatureImportanceTest.kt
package chapter10

import chapter03.DecisionTree
import chapter03.Leaf
import org.jetbrains.kotlinx.dataframe.api.dataFrameOf
import kotlin.test.Test
import kotlin.test.assertEquals

class TreeImportancesTest {
    @Test
    fun `分割しない木はすべての特徴量の重要度が0`() {
        val x =
            dataFrameOf(
                "がく片幅" to listOf(0.3, 0.5),
                "花弁幅" to listOf(0.1, 0.2),
            )
        val t = listOf("setosa", "setosa")

        assertEquals(mapOf("がく片幅" to 0.0, "花弁幅" to 0.0), treeImportances(Leaf(label = "setosa"), x, t))
    }

    @Test
    fun `1回だけ分割する木は分割に使った特徴量の重要度が1`() {
        val x =
            dataFrameOf(
                "がく片幅" to listOf(0.3, 0.5, 0.4, 0.6),
                "花弁幅" to listOf(0.1, 0.2, 0.8, 0.9),
            )
        val t = listOf("setosa", "setosa", "virginica", "virginica")
        val tree = checkNotNull(DecisionTree().fit(x, t).tree)

        assertEquals(mapOf("がく片幅" to 0.0, "花弁幅" to 1.0), treeImportances(tree, x, t))
    }

    @Test
    fun `分割で減った不純度を件数で重み付けして割合にする`() {
        val x =
            dataFrameOf(
                "花弁長さ" to listOf(0.1, 0.2, 0.3, 0.8, 0.7, 0.9),
                "花弁幅" to listOf(0.1, 0.1, 0.1, 0.2, 0.9, 0.9),
            )
        val t = listOf("setosa", "setosa", "setosa", "versicolor", "virginica", "virginica")
        val tree = checkNotNull(DecisionTree().fit(x, t).tree)

        val importances = treeImportances(tree, x, t)

        assertEquals(7.0 / 11, importances.getValue("花弁長さ"), absoluteTolerance = 1e-12)
        assertEquals(4.0 / 11, importances.getValue("花弁幅"), absoluteTolerance = 1e-12)
    }
}

class ForestImportancesTest {
    @Test
    fun `木が1本なら学習に使った行でのその木の重要度と一致する`() {
        val x =
            dataFrameOf(
                "がく片幅" to listOf(0.5, 0.3, 0.6, 0.4, 0.5, 0.3, 0.6, 0.4),
                "花弁幅" to listOf(0.1, 0.12, 0.14, 0.16, 0.8, 0.82, 0.84, 0.86),
            )
        val t = List(4) { "setosa" } + List(4) { "virginica" }
        val forest = RandomForest(nEstimators = 1, maxFeatures = 2, seed = 0).fit(x, t)
        val fitted = forest.trees.single()

        val expected = treeImportances(checkNotNull(fitted.model.tree), x[fitted.rows], t.slice(fitted.rows))

        assertEquals(expected, forestImportances(forest, x, t))
    }
}

File: apps/node/test/chapter10/feature-importance.test.ts
import { describe, expect, it } from "vitest";
import { DecisionTree, type Tree } from "../../src/chapter03/decision-tree.ts";
import {
  forestImportances,
  treeImportances,
} from "../../src/chapter10/feature-importance.ts";
import { RandomForest } from "../../src/chapter10/random-forest.ts";

function fitTree<K extends string>(
  x: readonly Record<K, number>[],
  t: readonly string[],
): Tree<K> {
  const tree = new DecisionTree<K>().fit(x, t).tree;
  if (tree === null) {
    throw new Error("決定木を学習できませんでした");
  }
  return tree;
}

describe("treeImportances", () => {
  it("分割しない木はすべての特徴量の重要度が 0", () => {
    const x = [
      { がく片幅: 0.3, 花弁幅: 0.1 },
      { がく片幅: 0.5, 花弁幅: 0.2 },
    ];
    const t = ["setosa", "setosa"];

    expect(treeImportances({ kind: "leaf", label: "setosa" }, x, t)).toEqual({
      がく片幅: 0,
      花弁幅: 0,
    });
  });

  it("1 回だけ分割する木は分割に使った特徴量の重要度が 1", () => {
    const x = [
      { がく片幅: 0.3, 花弁幅: 0.1 },
      { がく片幅: 0.5, 花弁幅: 0.2 },
      { がく片幅: 0.4, 花弁幅: 0.8 },
      { がく片幅: 0.6, 花弁幅: 0.9 },
    ];
    const t = ["setosa", "setosa", "virginica", "virginica"];
    expect(treeImportances(fitTree(x, t), x, t)).toEqual({
      がく片幅: 0,
      花弁幅: 1,
    });
  });

  it("分割で減った不純度を件数で重み付けして割合にする", () => {
    const x = [
      { 花弁長さ: 0.1, 花弁幅: 0.1 },
      { 花弁長さ: 0.2, 花弁幅: 0.1 },
      { 花弁長さ: 0.3, 花弁幅: 0.1 },
      { 花弁長さ: 0.8, 花弁幅: 0.2 },
      { 花弁長さ: 0.7, 花弁幅: 0.9 },
      { 花弁長さ: 0.9, 花弁幅: 0.9 },
    ];
    const t = [
      "setosa",
      "setosa",
      "setosa",
      "versicolor",
      "virginica",
      "virginica",
    ];
    const importances = treeImportances(fitTree(x, t), x, t);

    expect(importances.花弁長さ).toBeCloseTo(7 / 11, 12);
    expect(importances.花弁幅).toBeCloseTo(4 / 11, 12);
  });
});

describe("forestImportances", () => {
  it("木が 1 本なら学習に使った行でのその木の重要度と一致する", () => {
    const sepalWidths = [0.5, 0.3, 0.6, 0.4, 0.5, 0.3, 0.6, 0.4];
    const petalWidths = [0.1, 0.12, 0.14, 0.16, 0.8, 0.82, 0.84, 0.86];
    const x = sepalWidths.map((sepalWidth, i) => ({
      がく片幅: sepalWidth,
      花弁幅: petalWidths[i] as number,
    }));
    const t = [...Array(4).fill("setosa"), ...Array(4).fill("virginica")];
    const forest = new RandomForest({
      nEstimators: 1,
      maxFeatures: 2,
      seed: 0,
    }).fit(x, t);
    const [fitted] = forest.trees;
    if (fitted?.model.tree == null) {
      throw new Error("決定木を学習できませんでした");
    }

    const expected = treeImportances(
      fitted.model.tree,
      fitted.rows.map((row) => x[row] as (typeof x)[number]),
      fitted.rows.map((row) => t[row] as string),
    );

    expect(forestImportances(forest, x, t)).toEqual(expected);
  });
});
